Sources/Common/Communication/EventBusHandler.swift

```diff
@@ -1,66 +1,87 @@
-        subscribeAndReplay(eventType: eventType, action: action)
+    /// Adds an observer for a specific event type.
+    /// Replays any in-memory stored events of that type to the new observer.
+    public func addObserver<E: EventRepresentable>(_ eventType: E.Type, action: @escaping (E) -> Void) {
+        eventBus.addObserver(eventType, action: action)
+        replayEvents(forType: eventType)
     }
 
-    @discardableResult
-    public func onReceive<E: EventRepresentable, S: Scheduler>(_ eventType: E.Type, performOn scheduler: S, action: @escaping (E) -> Void) -> AnyCancellable {
-        subscribeAndReplay(eventType: eventType, scheduler: scheduler, action: action)
+    /// Replays events of a specific type from in-memory storage to the newly added observer.
+    /// Events that are successfully sent to an observer are removed from file-based storage.
+    private func replayEvents<E: EventRepresentable>(forType eventType: E.Type) {
+        let key = eventType.key
+        if let storedEvents = memoryStorage[key] as? [E] {
+            storedEvents.forEach {
+                let isSent = eventBus.post($0, on: nil)
+                if isSent {
+                    removeFromStorage($0)
+                }
+            }
+        }
     }
 
-    private func subscribeAndReplay<E: EventRepresentable>(
-        eventType: E.Type,
-        scheduler: (any Scheduler)? = nil, // Optional scheduler
-        action: @escaping (E) -> Void
-    ) -> AnyCancellable {
-        let listener = eventBus.listenersRegistry.getOrCreateListener(forEventType: E.self)
-
-        let subscription: AnyCancellable
-        if let scheduler = scheduler {
-            subscription = listener.registerSubscription(scheduler: scheduler, action: action)
-        } else {
-            subscription = listener.registerSubscription(action: action)
+    /// Posts an event to the EventBus.
+    /// If there are no observers, the event is stored in file-based storage so we don't lose it.
+    /// Regardless of being sent, the event is temporarily stored in in-memory storage for replay to new observers.
+    public func postEvent<E: EventRepresentable>(_ event: E) {
+        let hasObservers = eventBus.post(event, on: nil)
+        memoryStorage[event.key, default: []].append(event)
+        if !hasObservers {
+            storeEvent(event)
         }
+    }
 
-        let key = String(describing: E.self)
-
-        loadAndSendStoredEvents(ofType: E.self, to: listener)
-
-        return subscription
+    /// Stores an event in file-based persistent storage.
+    /// This method is used when an event is posted but no observers are currently registered.
+    private func storeEvent<E: EventRepresentable>(_ event: E) {
+        do {
+            try eventStorage.store(event: event)
+        } catch {
+            print("Error storing event: \(error)")
+        }
     }
```

Comment:
I could be wrong here, but I think that the persistent event storage is only needed for `delivered` push metrics, correct? 

If this is correct, I think that before we commit to this implementation, we first need to go to the team for a decision. A decision on how `delivered` push events are to be implemented in the CDP SDK at launch. 

I say this because the persistent storage makes this eventbus implementation more complex to test and maintain. If we are only adding this complexity for `delivered` metrics (which has multiple implementation paths available), that is something good to bring to the team.

Reply:
In the call, we did decide to persist the events, especially in the case of `track metric/in-app` because since push/NSE/wrapper have lifecycles of its own, we can't rely on that there can't be a case where if we store events just in the memory, the user can't kill the app and those events that stayed in memory can get lost forever. 

Hence, we need to make sure, all the events have at least been sent to 1 subscriber that was listening for those events.

Reply:
From my memory of the call, we agreed that a persistent event storage was needed only in the case of `delivered` events and a NSE. If we were to only talk about the host iOS app though, the only scenario where events could be lost is if the CIO SDK never gets initialized (which, I am not sure we need to support that use case?). 

If the above statement is true, then I do still believe that an in-memory store is all that we need in the eventbus.